### ISSUE TITLE
"documented symbol was not declared or defined" when definition is in a macro

### ISSUE THREAD

**Issue by dermojo** [doxygen/doxygen#7886]
**Describe the bug**
I'm using macros to declare and define boilerplate functions, including comments. Doxygen complains that the according function wasn't declared or defined.

**Expected behavior**
Expected the same documentation as for the non-macro version.

**To Reproduce**
Sample code with macro usage:
```cpp
/// helper macro
#define DECLARE_DESTR(Class) \
   /** destructor */ \
   ~Class() noexcept

/// another helper macro
#define DEFINE_DESTR(Class) \
   Class::~Class() noexcept = default;

/// my class
class C
{
public:
   DECLARE_DESTR(C);
};

DEFINE_DESTR(C)
```

Working version without macros:
```cpp
/// my class
class D
{
public:
   ~D() noexcept;
};

D::~D() noexcept = default;
```

Doxyfile:
```
ENABLE_PREPROCESSING = YES
MACRO_EXPANSION = YES
EXPAND_ONLY_PREDEF = NO
SKIP_FUNCTION_MACROS = NO
OUTPUT_DIRECTORY = out
WARN_LOGFILE = out/warn.log
GENERATE_LATEX = NO
```

I tested SKIP_FUNCTION_MACROS = NO and YES, no difference.

**Version**
1.8.19 (3e9d9386950cc67d155a920cf8483c4d5bdcec3c), Ubuntu 18.04, 64 bit, built from git myself (gcc 7)

**Comment by albert-github** [comment]
Looking with the preprocessor output we see that:
```
00001
00002 /// helper macro
00003 #define DECLARE_DESTR(Class)
00004
00005
00006
00007 /// another helper macro
00008 #define DEFINE_DESTR(Class)
00009
00010
00011 /// my class
00012 class C
00013 {
00014 public:
00015    /** destructor */   ~ C () noexcept;
00016 };
00017
00018  C ::~ C () noexcept = default;
```
creating an equivalent version without macros:
```
/// my class
class E
{
public:
   /** destructor */
   ~ E () noexcept;
};

 E ::~ E () noexcept = default;
```
We also get here the message:
```
warning: documented symbol 'E::~E' was not declared or defined.
```
it looks like doxygen doesn't like the space in `~ E () noexcept;`

I don't know, by head, whether or not a space is allowed directly after the `~` (also of interest for `E`).

**Comment by dermojo** [comment]
I consulted my compiler and it seems that a space after the tilde is fine.
However, the preprocessor doesn't need to add one.

**Comment by albert-github** [comment]
>I consulted my compiler and it seems that a space after the tilde is fine.

   some compilers are a little bit less stringent as a standard. Did you try to compile with options like `-Wpedantic` or `-std=...` (e.g. `-std=c++14`) as well

> However, the preprocessor doesn't need to add one. 

true (but there must have been a reason in the doxygen preprocessor for adding it).

**Comment by dermojo** [comment]
> some compilers are a little bit less stringent as the cpp standard is. Did you try to compile with options like -Wpedantic or -std=... (e.g. -std=c++14) as well

I re-checked and tried with ``g++ -Wall -Wextra -std=c++17`` as well as ``clang++-10 -Weverything``. Looks fine, although I have never seen any code that uses it...

> true (but there must have been a reason in the doxygen preprocessor for adding it).

Sure, and my guess is that it's somehow related to the ``Preprocessor``'s ``nospaces``. I looked into it, but have no clue where to start.

**Comment by albert-github** [comment]
@doxygen 

It looks like the following patch in `pre.l` just above `<DefineText,CopyCComment>{ID}                {` fixes the issue:
```
<DefineText,CopyCComment>~{ID}                {
                                          yyextra->nospaces=TRUE;
                                          REJECT;
                                        }
```

- I'm a bit hesitant for possible side effects though
- shoud a construct with a space like `~ E` be supported in `scanner.l` as well?

**Comment by doxygen** [comment]
@albert-github yes it is just valid C++, so it should be supported by scanner.l (and then you probably don't even need to change pre.l).

**Comment by albert-github** [comment]
@doxygen Thanks.

I'll have a go at scanner.l
When looking at the output (with the space due to the preprocessing), the space is visible in e.g. html output:

![with_space](https://user-images.githubusercontent.com/5223533/86128748-13748500-bae2-11ea-9495-89f3a42451f9.jpg)

and xml output: `<definition>C::~ C</definition>`

- wouldn't this justify a, separate, patch for the space in `pre.l`?

**Comment by albert-github** [comment]
I did (I think) identify the rule that might have to be changed (in `scanner.l`):
```
<ReadBody,ReadNSBody,ReadBodyIntf>[^\r\n\#{}"@'/<]*     { yyextra->current->program += yytext ; }
```
but this rule is not only used for functions etc but also for other text. Using `removeRedundantWhiteSpace()` does not work for this reason and also it does not see `~` as something to remove the spaces as well.
Another complicating factor is that also in the source code (links and other documentation links) the `~ ` type should be handled.

At the moment I don't see a solution for this general problem in `scanner.l` (and `code.l`).
